libfwupdplugin/fu-cpd-firmware.c

```diff
@@ -0,0 +1,426 @@
+/*
+ * Copyright (C) 2022 Richard Hughes <richard@hughsie.com>
+ * Copyright (C) 2022 Intel
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN "FuFirmware"
+
+#include "config.h"
+
+#include <string.h>
+
+#include "fu-byte-array.h"
+#include "fu-bytes.h"
+#include "fu-cpd-firmware.h"
+#include "fu-mem.h"
+#include "fu-string.h"
+
+/**
+ * FuCpdFirmware:
+ *
+ * A CPD often found in PCI OptionROM firmware images.
+ *
+ * See also: [class@FuFirmware]
+ */
+
+typedef struct {
+	guint8 header_version;
+	guint8 entry_version;
+} FuCpdFirmwarePrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE(FuCpdFirmware, fu_cpd_firmware, FU_TYPE_FIRMWARE)
+#define GET_PRIVATE(o) (fu_cpd_firmware_get_instance_private(o))
+
+#define FU_CPD_FIRMWARE_HEADER_MARKER 0x44504324
+
+typedef struct __attribute__((packed)) {
+	guint32 header_marker;
+	guint32 num_of_entries;
+	guint8 header_version;
+	guint8 entry_version;
+	guint8 header_length;
+	guint8 checksum;
+	guint32 partition_name;
+	guint32 crc32;
+} FuCpdDirectoryHeader;
+
+typedef struct __attribute__((packed)) {
+	gchar name[12];
+	guint32 offset;
+	guint32 length;
+	guint8 reserved1[4];
+} FuCpdDirectoryEntry;
+
+typedef struct __attribute__((packed)) {
+	guint32 header_type;
+	guint32 header_length;
+	guint32 header_version;
+	guint32 flags;
+	guint32 vendor;
+	guint32 date;
+	guint32 size;
+	guint32 id;
+	guint32 rsvd;
+	guint64 version;
+	guint32 svn;
+	guint64 rsvd1;
```

Comment:
if this and the next one are both reserved, why have separate variables?

Reply:
Agreed; we need neither, both removed.